src/Sylius/Bundle/CoreBundle/EventListener/OrderCurrencyListener.php

```diff
@@ -48,6 +63,12 @@ public function processOrderCurrency(GenericEvent $event)
             );
         }
 
+        $currency = $this->currencyRepository->findOneBy(['code' => $this->currencyContext->getCurrency()]);
+
+        if ($this->currencyContext->getDefaultCurrency() !== $exchangeRate = $currency->getCode()) {
+            $order->setExchangeRate($currency->getExchangeRate());
+        }
```

Comment:
You can save yourself a possible repository call here.

Check to see `if ($order->getCurrency() != $this->currencyContext->getCurrency()) {` and only then look it up from the repo to se the exchange rate. 

It's possible that you could update the order less frequently, but I think this is probably preferable. 

However I'm now starting to wonder if this is the wrong place to set it... Maybe it would be better set right at the time of order completion? Since that will be the most accurate time to do it.

Can you find when the order gets completed just to see if it's easy to do?